### ISSUE TITLE
If requesting client is located on a registered xrootd server then redirect to that server

### ISSUE THREAD

**Issue by lickdragon** [xrootd/xrootd#1306]
Hello all,
In our cluster the amount of network usage could be reduced by half with a change to the redirector selection algorithm.

In our cluster:
- Every node runs Condor and xrootd server:
`┌───────┐ ┌───────┐ 
│server 1│ │server 2│`
`
│             │ │             │  .....
│ condor │ │ condor │
│             │ │             │
│ xrootd  │ │ xrootd  │
└───────┘ └───────┘
`
- There is an xrootd redirector that all the servers are registered with.
- Condor starts a job, the job requests files from the xrootd redirector.
- The redirector chooses an xrootd server "at random" among least loaded servers. 
- Most of the time, the xrootd server chosen is not on the same server as the requesting job, leading to a transfer of the file from the chosen server to the client server.  The file is sent over the network twice:

                             ┌───────┐   ┌───────┐ 
                             │server 1│    │server 2│
storage node ─┐   │             │    │             │  .....
storage node ─┼┐  │condor  │┌─┼ condor │
storage node ─┘│  │             ││  │             │
                          └─┼xrootd   ┼┘  │ xrootd  │
                              └───────┘    └───────┘

- If the redirector would notice that client request came from a registered xrootd server, it could chose the same server to serve the file.  The file would be sent over the network once:

                              ┌───────┐ 
                              │server 2│
storage node ─┐    │             │
storage node ─┼┐  │ condor │
storage node ─┘│  │      │     │
                          └─┼ xrootd  │
                              └───────┘

The overall idea is similar to the XrdCmsRedirLocal xrootd plugin.  One difference is that the xrootd server will not necessarily be using the file:// protocol.  (Instead the xrootd server uses an xrootd/HDFS ofs.osslib plugin.)  The second possible difference is that the XrdCmsRedirLocal redirects to local if client IP and target IP are both "private".  What does private means?  For there to be network savings in our situation, the xrootd client IP should equal one of the registered  server IPs (and the client is then redirected to that server).

Possibly "if requesting client is located on a registered xrootd server then redirect to that server" would be generally useful heuristic?  At the moment I cannot think of a scenario where it would be inefficient or undesired.

Thanks for your work!
Chad.

**Comment by abh3** [comment]
Could you cut a ghithub ticket on this. We recognize this is a desirable 
enhancement but we need a ticket to track it.

Andy

**Comment by lickdragon** [comment]
Hi Andy,
   Opened one here:
   https://github.com/xrootd/xrootd/issues/1306

Thanks for your hard work!
Chad.

**Comment by abh3** [comment]
While we technically agree with your request (i.e. that should be the case) the solution is far more complicated than one would imagine. The reason is that the server in question may be fronted by a supervisor if you happen to have more than 64 nodes (which I think you do). So, matching incoming request with same-node server existence is rather complicated. So, unfortunately, we need to put this in the queue as it will not be a straightforward change.

**Comment by lickdragon** [comment]
Hi Andrew,
   Thanks for taking a look.  As you guessed we do have supervisors. 
Are supervisors an absolute requirement for over 64 or can a fast enough 
redirector handle more?
   Also, I don't know enough about xrootd to answer this question: Are 
requests divided amongst the redirector and the supervisors?  If so, and 
the redirector had this new heuristic, some of the file requests would 
be served by same-node.  Something is better than nothing!

Thanks again!
Chad.

**Comment by xrootd-dev** [comment]
Hi Chad,

No, 64 is the hard limit for any server. You may wonder why. We made sure 
that only constant time algorithms were used for any lookup processing. 
That effectively put a limit on how many nodes could be connected to a 
server so as to not have the constant time overhead as a dominant factor 
when nodes were added. So, the solution is the true and tried divide and 
conquer technique. By arranging lookups into a B-Tree we could make sure 
that (where k is fixed overhead and n is the number of nodes) k*n 
processing time rapidly decreases relative to K*log64(n) as n increases. 
This means as you add more nodes perfromace doesn't detrioate.

But why 64, wouldn't larger log values work better? Good question. It all 
comes down to the hardware. A 64 bit vector size is the most efficient
size for today's computers. Since most of the lookup processing involves 
vector bit slicing; a 64 bit vector is the most efficient way of doing bit 
slicing on a CPU today. Yes, GPU processors give you much larger vectors 
sizes and some of collaborators have suggested going down the GPU path. 
Unfortunately, the GPU environment is unpredictable so we decided not to 
go down that path.

Of course, by arranging lookups in a distributed B64-tree lookup we 
sacrificed specificity. It was a reasonable approach at the time to make 
lookups as fast as possible. Today's requirements are in conflict with 
that decision. You are not the only one that suffers from the lack of 
specificity.

All of that said, we are looking at ways to address this conflict. The 
outlook is positive. We were very much constrained by having to support 
RH6 platforms with very primitive atomics support. So, once we can drop 
that support it would allow us to reduce the "k" which allows us to 
increase the size of the bit vector and still maintain good performance. 
That may not be sufficient for your site but we are looking at all the 
alternatives.

I know I gave you far more informatuon than you wanted. However, it's 
important to understand that constraints are not chosen at random or 
for convenience. We do try to optimize the sevrer for the common use 
cases. But, we also recognize those will change and we will have to 
readjust our optimization targets. That's the process we are in today, so 
bear with us.

Andy

**Comment by lickdragon** [comment]
Wow!  I would never have guessed xrootd needed such optimizations! 
Happily, it looks like EL6 (CentOS 6) won't be getting updates anymore 
after Nov 30th.  I suppose you all have got a lot of changes planned for 
the its death-day party. :)

Thanks for all your expertise,
Chad.

**Comment by xrootd-dev** [comment]
Indeed, it will be good day. The xroot side of support ends with release 
5.1.0 out in a couple of weeks, if not sooner. Existing releases get 
support for SL6 until the end of March. Now, it would be good if you send 
me your config files (i.e. the ones you use on the local workers and the 
redirector). I wouldn't suggest you post them unless they are fully 
public. Also, how many worker nodes are we talking about?

Andy

**Comment by lickdragon** [comment]
> Also, how many worker nodes are we talking about?
Er, well, due to historical choices, about 600 ATM. :/  (We 
hyperconverged before it was a buzzword. :) )

> Now, it would be good if you send me your config files (i.e. the ones
> you use on the local workers and the redirector). I wouldn't suggest
> you post them unless they are fully public. 

Is vaguely like this your email abh@SLAC._______.EDU ?
I don't think there is anything secret in our config, but possibly clueless.

Chad.

**Comment by xrootd-dev** [comment]
On Tue, 17 Nov 2020, cwseys wrote:

> Is vaguely like this your email abh@SLAC._______.EDU ?
Yep, just fill in the obvious blank and you have it.

Andy

**Comment by abh3** [comment]
@cwseys Are you still running this kind of setup? I think we have a plan that would solve the issue without a lot of extra code.

**Comment by abh3** [comment]
Good to know. I think I can solve this is a relatively short amount of time. Not sure whether it will hit 5.7.1 but will certainly be included in 6.0.

**Comment by abh3** [comment]
One more thing. Just so that the right thing gets implemented. When a client requests a file but the file is not present in he server on the client's node, the client should be redirecte to where the file resides. If the file is on he client's node then redirect to that node. Now, if the fie is not present anywhere and this is for a "create" request, always redirect the client to the server local to the client. Right?

**Comment by cwseys** [comment]
> Right?
Yes, in our particular setup every xrootd file server has access to the same file space.  Writing the file by communicating with the local xrootd server would be ideal.

Just to be clear, in our situation there are some condor nodes which are not also xrootd file servers.  They would have to talk to a different node.  (I wouldn't want you to spend time on code that works for a mis-communicated scenario!)

Chad.

**Comment by abh3** [comment]
So we are back to the issue of 64. How many nodes do you have? Is this a read/only setup (i.e. we don't have to worry about write/create requests)? Also, tell me what you cms.dfs directive is.

**Comment by cwseys** [comment]
> So we are back to the issue of 64. How many nodes do you have? Is this a read/only setup (i.e. we don't have to worry about write/create requests)? Also, tell me what you cms.dfs directive is.

Yeah, we have more than 64 xrootd file servers (and we have xrootd 
supervisors as a result).  File creation and writing is also allowed. :)

# grep cms.dfs /etc/xrootd/xrootd-t2wisc.cfg
cms.dfs lookup distrib mdhold 60s redirect immed

Thanks!
Chad.
